Add tests involving destroying Katana objects after thread pool shutdown

arthurp opened the pull request:
This can happen pretty easily from Python, so it's worth keeping track of. This stops working, we could probably figure out a way to avoid this problem and no longer need these tests.

witchel commented on libsupport/src/Uri.cpp:
```diff
@@ -18,12 +19,54 @@ namespace {
+    'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm',
+    'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z',
+    '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '-', '_'};
+
```
Should we use this standard?
https://tools.ietf.org/html/rfc4648#section-5
It seems nearly identical to what you did.  Seems like a standardized safe URL format.
https://github.com/tplgy/cppcodec

tylershunt replied:
I think this code does meet this standard, or did I mix up some characters? Since this was pretty simple I thought a code fragment copied from libstanford (or whatever they call it) was better than adding a dependency. I could be convinced the other way though.

witchel replied:
Yeah, I can see that.  I think there is `=` padding that I didn't verify in your implementation.

ddn0 replied:
if it's voting time, I'd put my shekels on just copying the functionality since a correct implementation isn't that much work and the effort of dependency management seems like a lot just for this one capability.

tylershunt replied:
> Yeah, I can see that. I think there is `=` padding that I didn't verify in your implementation.

Its slightly convoluted but since the implementation I borrowed allocates a string that is an acceptable length, I initialize that string to all `=` so the padding should be handled if the string itself is an odd length.

FWIW I've verified that the golang's standard base 64 implementation decodes this correctly.

witchel replied:
Sweet.

tylershunt replied:
(I should also note that this was merged in another PR that this work happened to depend on, but I'm happy to fix/change this to whatever we decide to do here)

tylershunt replied:
I think we have consensus to leave this, if i'm wrong let me know.

witchel replied:
100%

witchel commented on libsupport/src/Uri.cpp:
```diff
@@ -18,12 +19,54 @@ namespace {
-    : scheme_(std::move(scheme)), path_(std::move(path)) {
+    : scheme_(std::move(scheme)),
+      path_(std::move(path)),
+      string_(scheme_ + "://" + path_) {
```
Why do you use `std::move` for the string?

ddn0 replied:
There are a couple ways to initialize an object with a copy of an argument:

1. take const reference then apply copy constructor: `A(const T& t): t_(t) {}`
2. copy in argument then apply move constructor: A(T t): t_(std::move(t)) {}`

In both cases, a copy is made, but the second method also supports cases like `A(std::move(o))` with no additional code at the expense of an extra move in all cases.

Here's some more context: https://stackoverflow.com/questions/21035417/is-the-pass-by-value-and-then-move-construct-a-bad-idiom

witchel replied:
Interesting article.  And a string is constant-time copyable, but we do have PODs in our code base, so I guess the move idiom is not best in those cases?

ddn0 replied:
Correct.

witchel commented on libsupport/src/Uri.cpp:
```diff
@@ -51,6 +94,20 @@ Uri::Make(const std::string& str) {
+  }
+  return false;
+}
+
```
So s3:// is not empty?  I guess not.  But it is kinda.

tylershunt replied:
I think `s3://` would not be possible because of the regex sanitization that I do in `make` that was my intent anyway. I'll double check. (This was also true in the last PR I just moved this def here while I was cleaning up `Uri.h`)

tylershunt replied:
Verified, in the regex I check for `(.+)` on the right side of `//` so a `Uri` object for only `s3://` should be impossible to construct.

witchel commented on libtsuba/include/tsuba/Errors.h:
```diff
@@ -67,6 +68,8 @@ class ErrorCodeCategory : public std::error_category {
       return "previous version expectation violated";
+    case ErrorCode::MpiError:
+      return "some mpi process reported an error";
     default:
```
In a reported error, shouldn't we spell it MPI?

tylershunt replied:
Ah yeah good catch.

witchel commented on libtsuba/src/GlobalState.h:
```diff
@@ -63,6 +64,24 @@ galois::CommBackend* Comm();
 NameServerClient* NS();
 
+/// Execute a call back only on one host, return it's return value to all hosts
+template<typename Callable>
```
I read that comment three times.  Third time was the charm.

witchel commented on libtsuba/src/RDG.cpp:
```diff
@@ -736,7 +735,7 @@ CommitRDG(
-  if (comm->ID == 0) {
+  galois::Result<void> ret = tsuba::OneHostOnly([&]() -> galois::Result<void> {
     TSUBA_PTP(tsuba::internal::FaultSensitivity::High);
 
```
Very cool that this code will allow me to return an error in the body of this function, while right now I just have to print an error.

tylershunt replied:
Yeah, I think returning errors is harder to get wrong too.

witchel reviewed (approved):
I like the MPI wrapper, very cool.

ddn0 commented on libdist/src/NetworkBuffered.cpp:
```diff
@@ -329,8 +330,17 @@ class NetworkInterfaceBuffered : public NetworkInterface {
-  virtual void NotifyFailure() { MPI_Abort(MPI_COMM_WORLD, 1); }
+  void Barrier() override { galois::runtime::getHostBarrier().wait(); }
+
+  void Broadcast(bool* val, uint32_t root) override {
```
more of a discussion point, but would we rather have a more C++ like interface rather than a straight port of the MPI call? E.g.,
```
bool Broadcast(uint32_t root, bool val)
```

tylershunt replied:
Yeah I went back and forth on this; I think hearing someone else say that we should have a more c++ interface is enough to convince me.

ddn0 commented on libtsuba/src/GlobalState.h:
```diff
@@ -63,6 +64,24 @@ galois::CommBackend* Comm();
 
+/// Execute a call back only on one host, return it's return value to all hosts
+template<typename Callable>
+galois::Result<void> OneHostOnly(Callable cb) {
```
Perhaps take a https://en.cppreference.com/w/cpp/utility/functional/function instead and we can avoid templatization at the expense of a virtual function call.

witchel replied:
I wondered about this.  The tradeoff you say makes sense.  This is more generic advice that doesn't apply to us as much.

https://stackoverflow.com/questions/14677997/stdfunction-vs-template#:~:text=std%3A%3Afunction%20is%20not,is%20unknown%20at%20compile%2Dtime.

tylershunt replied:
Yeah I didn't go this way initially because `std::function` has some weird restrictions (you can't use the new lambdas that capture by move for example), but I'm not doing that anyway and I think capture by reference is probably always appropriate here since this isn't async or anything like that.

roshandathathri commented on libdist/include/galois/runtime/Network.h:
```diff
@@ -153,8 +153,9 @@ class GALOIS_EXPORT NetworkInterface : public CommBackend {
-  virtual void NotifyFailure() = 0;
+  void Barrier() override = 0;
+  void NotifyFailure() override = 0;
+  bool Broadcast(uint32_t root, bool val) override = 0;
```
Do you want this to be general? If so, why not use SendBuffer instead of val?

Also, there are a couple of broadcast* functions in this class already and neither of them are used currently. Maybe just delete them to avoid confusion?

tylershunt replied:
I'd rather keep this simple and complicate/generalize later as needed. I can clean up the other `broadcast` functions, I'll do that.

roshandathathri replied:
Makes sense

roshandathathri commented on libtsuba/tests/MPICommBackend.h:
```diff
@@ -53,6 +53,14 @@ struct MPICommBackend : public galois::CommBackend {
+    if (MPI_Bcast(&val_buf, 1, MPI_INT, root, MPI_COMM_WORLD) != 0) {
+      GALOIS_LOG_FATAL("MPI_Bcast failed");
+    }
+    *val = static_cast<bool>(val_buf);
```
Return type is different for this?

roshandathathri reviewed (approved):
LGTM

roshandathathri commented on libdist/include/galois/runtime/Network.h:
```diff
@@ -151,10 +140,6 @@ class GALOIS_EXPORT NetworkInterface : public CommBackend {
       const = 0;
-
-  ///// Required CommBackend overrides
-  virtual void Barrier() = 0;
```
Why were these two functions removed?

witchel replied:
I don't want to speak for Tyler, but I think with his `tsuba::OneHostOnly`, there is no need for an explicit call to Barrier, which is less error prone.

arthurp committed:
Add tests involving destroying Katana objects after thread pool shutdown.

This can happen pretty easily from Python, so it's worth keeping
track of. This stops working, we could probably figure out a way
to avoid this problem and nolonger need these tests.

arthurp committed:
Fix use of xfail.

roshandathathri reviewed (approved):
LGTM. Thanks!
